src/Custom/Common/Instrumentation/InstrumentationScope.cs

```diff
@@ -0,0 +1,218 @@
+    }
+
+    public static InstrumentationScope StartChat(string model, string operationName,
+        string serverAddress, int serverPort, ChatCompletionOptions options)
+    {
+        if (s_chatSource.HasListeners() || s_tokens.Enabled || s_duration.Enabled)
+        {
+            var scope = new InstrumentationScope(model, operationName, serverAddress, serverPort);
+            scope.StartChat(options);
+            return scope;
+        }
+
+        return null;
+    }
+
+    private void StartChat(ChatCompletionOptions options)
+    {
+        _duration = Stopwatch.StartNew();
+        _commonTags = new TagList
+        {
+            { Constants.GenAiSystemKey, Constants.GenAiSystemValue },
+            { Constants.GenAiRequestModelKey, _requestModel },
+            { Constants.ServerAddressKey, _serverAddress },
+            { Constants.ServerPortKey, _serverPort },
+            { Constants.GenAiOperationNameKey, _operationName },
+        };
+
+        _activity = s_chatSource.StartActivity(string.Concat(_operationName, " ", _requestModel), ActivityKind.Client);
+        if (_activity?.IsAllDataRequested == true)
+        {
+            RecordCommonAttributes();
+            SetActivityTagIfNotNull(Constants.GenAiRequestMaxTokensKey, options?.MaxTokens);
+            SetActivityTagIfNotNull(Constants.GenAiRequestTemperatureKey, options?.Temperature);
+            SetActivityTagIfNotNull(Constants.GenAiRequestTopPKey, options?.TopP);
+        }
+
+        return;
+    }
+
+    public void RecordChatCompletion(ChatCompletion completion)
+    {
+        RecordMetrics(completion.Model, null, completion.Usage?.InputTokens, completion.Usage?.OutputTokens);
+
+        if (_activity?.IsAllDataRequested == true)
+        {
+            RecordResponseAttributes(completion.Id, completion.Model, completion.FinishReason, completion.Usage);
+        }
+    }
+
+    public void RecordException(Exception ex)
+    {
+        var errorType = GetErrorType(ex);
+        RecordMetrics(null, errorType, null, null);
+        SetActivityError(ex, errorType);
+    }
+
+    public void Dispose()
+    {
+        _activity?.Stop();
+    }
+
+    private void RecordCommonAttributes()
+    {
+        _activity.SetTag(Constants.GenAiSystemKey, Constants.GenAiSystemValue);
+        _activity.SetTag(Constants.GenAiRequestModelKey, _requestModel);
+        _activity.SetTag(Constants.ServerAddressKey, _serverAddress);
+        _activity.SetTag(Constants.ServerPortKey, _serverPort);
+        _activity.SetTag(Constants.GenAiOperationNameKey, _operationName);
+    }
+
+    private void RecordMetrics(string responseModel, string errorType, int? inputTokensUsage, int? outputTokensUsage)
+    {
+        TagList tags = ResponseTagsWithError(responseModel, errorType);
+        s_duration.Record(_duration.Elapsed.TotalSeconds, tags);
+
+        if (inputTokensUsage != null)
+        {
+            // tags is a struct, let's copy them
+            TagList inputUsageTags = tags;
```

Comment:
Token usage metrics has no attribute for error.type, but tags here will have error.type.

Reply:
it should be impossible to have an error AND usage data (it comes from response body), but I refactored it a bit to be more obvious